service/dap/server_test.go

```diff
@@ -1709,33 +1689,48 @@ func TestAttachSubstitutePath(t *testing.T) {
 		t.Skip("test skipped on windows, see https://delve.beta.teamcity.com/project/Delve_windows for details")
 	}
 	runTest(t, "loopprog", func(client *daptest.Client, fixture protest.Fixture) {
-		localDirPath := filepath.Join(string(filepath.Separator), "path", "that", "does", "not", "exist")
-		if runtime.GOOS == "windows" {
-			localDirPath = "C:" + localDirPath
-		}
-		localFilePath := filepath.Join(localDirPath, "loopprog.go")
-		debuggerDirPath := filepath.Dir(fixture.Source)
-
-		// Start the program to attach to
 		cmd := execFixture(t, fixture)
 
-		runDebugSessionWithBPs(t, client, "attach",
-			// Attach
-			func() {
-				client.AttachRequest(map[string]interface{}{"mode": "local", "processId": cmd.Process.Pid, "stopOnEntry": false, "substitutePath": []map[string]string{{"from": localDirPath, "to": debuggerDirPath}}})
-			},
-			// Set breakpoints
-			localFilePath, []int{8},
-			[]onBreakpoint{{
-				// Stop at line 8
-				execute: func() {
-					handleStop(t, client, 1, "main.loop", 8)
-				},
-				disconnect: true,
-			}})
+		substitutePathTestHelper(t, fixture, client, "attach", map[string]interface{}{"mode": "local", "processId": cmd.Process.Pid})
 	})
 }
 
+func substitutePathTestHelper(t *testing.T, fixture protest.Fixture, client *daptest.Client, request string, launchAttachConfig map[string]interface{}) {
+	t.Helper()
+	nonexistentDir := filepath.Join(string(filepath.Separator), "path", "that", "does", "not", "exist")
+	if runtime.GOOS == "windows" {
+		nonexistentDir = "C:" + nonexistentDir
+	}
+
+	launchAttachConfig["stopOnEntry"] = false
+	launchAttachConfig["substitutePath"] = []map[string]string{
+		{"from": nonexistentDir, "to": filepath.Dir(fixture.Source)},
+		{"from": filepath.Dir(fixture.Source), "to": filepath.Dir(fixture.Source)}, // Since the path mappings are ordered, this should never apply.
```

Comment:
Ok I updated them with some more description. Let me know if thats any clearer.

Reply:
> Ok I updated them with some more description. Let me know if thats any clearer.

Not really, I think they are just saying in more words the same thing as before. But upon further thinking, I think I can articulate what was confusing to me.

I was confused because "from" and "to" were identical, so it did not seem like there was anything to apply. Plus it was the only rule with "from" of that value, so did this have to do with order?

But now that I think about this more, I realize that the mapping will be applied in the direction "from"=>"to" for setting a breakpoint, something helper explicitly tests for, AND in the direction "to" => "from" for stacktrace, which is something the helper does not explicitly test for, so I forgot it was even there. It would be nice to be explicit about both directions in the test. Otherwise, it really is not clear what you mean by "applied".

Finally, this made me realize how incredibly confusing "from" and "to" labels are because they make you think about only one direction in a bi-directional relationship. Can we use "client" and "server" in the config as well?

Reply:
I agree that it is confusing. I'll try to make it less confusing, by changing the target mapping to a different path.

From offline discussion, we are not set on client and server as this may be confusing if users use this for different things. I think for now, we should leave it as "from" and "to" in order to stay consistent with dlv cli, and we can continue this discussion separately.